### ISSUE TITLE
Data Leakage

### ISSUE THREAD

**Issue by RingoTC** [zyy-2001/DisKT#2]
https://github.com/zyy-2001/DisKT/blob/master/diskt/models/diskt.py#L213

Performing softmax twice will cause the attention matrix to be redistributed, making the mask invalid. This fixed version:

```python
def contradictory_attention(query, key, value1, value2, mask=None, dropout=None, counter_attention_mask=None):
    "Compute 'Scaled Dot Product Attention'"
    bs, head, seqlen, d_k = query.size(0), query.size(1), query.size(2), query.size(-1)
    device = query.device
    
    scores = torch.matmul(query, key.transpose(-2, -1)) / math.sqrt(d_k)
    
    if mask is not None:
        scores = scores.masked_fill(mask == 0, -1e32)
    
    p_attn = F.softmax(scores, dim=-1)  # [batch_size, head, seq_len, seq_len]
    
    counter_attention_mask = counter_attention_mask.unsqueeze(1)
    counter_attention_mask = counter_attention_mask.expand(-1, head*seqlen, -1)
    counter_attention_mask = counter_attention_mask.reshape(-1, seqlen)
    
    attn_reshape = p_attn.reshape(bs*head*seqlen, -1)
    
    masked_attn = torch.where((counter_attention_mask == 1), 
                             torch.zeros_like(attn_reshape), 
                             attn_reshape)
    
    masked_attn = masked_attn.reshape(bs, head, seqlen, -1)
    
    row_sums = masked_attn.sum(dim=-1, keepdim=True)
    
    valid_rows = (row_sums > 0).float()
    safe_row_sums = row_sums + (1 - valid_rows)
    
    normalized_attn = masked_attn / safe_row_sums
    
    pad_zero = torch.zeros(bs, head, 1, seqlen).to(device)
    normalized_attn = torch.cat([pad_zero, normalized_attn[:, :, 1:, :]], dim=2)
    
    if dropout is not None:
        normalized_attn = dropout(normalized_attn)
    
    output_v1 = torch.matmul(normalized_attn, value1)
    output_v2 = torch.matmul(normalized_attn, value2)
    
    return output_v1, output_v2, normalized_attn
```

**Comment by zyy-2001** [comment]
We are very grateful for you pointing out the issue of data leakage, which is a common mistake in KT. This may have been overlooked due to our early stopping strategy. We have now fixed the relevant code based on the version you provided. It is worth noting that our re-softmax approach was inspired by sparseKT [1], an excellent piece of work, which might also suffer from the same issue.

As researchers, we should identify such mistakes and propose solutions. However, our current work does not focus on KT. We are also aware that you or your team are committed to building a unified and concise KT framework and have established a related leaderboard.

We sincerely ask you to reproduce our experimental results. We remain confident in the corrected method and will update our results based on your data.

Thank you very much for your correction.

[1] Towards Robust Knowledge Tracing Models via k-Sparse Attention

**Comment by zyy-2001** [comment]
In addition, we recall that the AKT[1] model implemented in CL4KT [2] exhibited potential data leakage issues on the spanish dataset, which we had identified earlier. However, the AKT model implemented within the pyKT framework [3] did not present such problems. It is worth noting that our previous work [4] categorized this type of issue as a problem of inapplicability and proposed corresponding solutions.

[1] Context-Aware Attentive Knowledge Tracing
[2] Contrastive Learning for Knowledge Tracing
[3] pyKT: A Python Library to Benchmark Deep Learning based Knowledge Tracing Models
[4] Revisiting Applicable and Comprehensive Knowledge Tracing in Large-Scale Data

**Comment by RingoTC** [comment]
Thanks for you kindly reply. Yes, I rerun some experiments of my fixed version based on your codebase, and in many case DisKT can also outperforms other baselines.

**Comment by zyy-2001** [comment]
Thank you for reproducing the results—We are really looking forward to your work!

**Comment by zyy-2001** [comment]
Thank you for your continued attention and support. We noticed that in the modified version of DisKT, the softmax operation was not applied, which appears to be an unreasonable approach. After re-evaluating the double softmax process, we found that masking with very small values and then applying softmax still results in extremely small outputs. This behavior aligns with the approach used in sparseKT and seems to be a reasonable design choice.

We believe that if there were truly a data leakage issue, the model’s performance would inflate prematurely across all datasets. Moreover, based on our recollection, the model has demonstrated consistently strong performance on certain datasets, such as dbe-kt22. Therefore, we have decided to revert to the original implementation.

If you make any new discoveries or have further insights, please feel free to share them with us at any time.
